File: Year-2022/LeetCode-December-Challenge/KeysandRooms.java
import java.util.*;
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        Queue<Integer> q = new LinkedList<>();
        boolean[] visited = new boolean[rooms.size()];
        q.add(0); visited[0] = true;
        while(!q.isEmpty()){
            int curr = q.poll();
            for(int x: rooms.get(curr)){
                if(!visited[x]){
                    visited[x] = true;
                    q.add(x);
                }
            }
        }
        for(boolean x: visited) if(!x) return false;
        return true;
    }
}
public class KeysandRooms {
    public static void main(String[] args) {
    }
}

File: Year-2022/LeetCode-December-Challenge/KeysandRooms.js
var canVisitAllRooms = function (rooms) {
    let q = [], visited = { '0': true };
    q.push(0), visited[0] = true;
    while (q.length) {
        let curr = q.pop();
        for (let x of rooms[curr]) {
            if (!visited[x]) {
                visited[x] = true;
                q.push(x);
            }
        }
    }
    return Object.keys(visited).length === rooms.length;
};
let rooms = [[1, 3], [3, 0, 1], [2], [0]];
console.log(canVisitAllRooms(rooms));